OAK-9004: Multitenant index name strategy with reindex support

averma21 opened the pull request.

Commit:
OAK-9004: Multitenant index name strategy with reindex support

Commit:
OAK-9004: Multitenant index name strategy with reindex support
* Removing unintended logs and fixing name

nit0906 commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderService.java:
```diff
@@ -140,6 +149,8 @@ private void activate(BundleContext bundleContext, Map<String, ?> config) {
 
         elasticsearchConnection = getElasticsearchCoordinate(config);
+        indexPrefix = PropertiesUtil.toString(config.get(PROP_INDEX_PREFIX), PROP_INDEX_PREFIX_DEFAULT);
+        LOG.info("Got indexPrefix=" + indexPrefix);
```
Can we change this to something like LOG.info("IndexPrefix - {} configured for Elastic Search indices", indexPrefix)

averma21 replied:
I actually deleted it as I didn't intend to add it. May be I can re add it. what about level - Info or debug?

nit0906 replied:
debug should be ok I suppose

nit0906 commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexProviderService.java:
```diff
@@ -140,6 +149,8 @@ private void activate(BundleContext bundleContext, Map<String, ?> config) {
         //initializeExtractedTextCache(config, statisticsProvider);
 
         elasticsearchConnection = getElasticsearchCoordinate(config);
+        indexPrefix = PropertiesUtil.toString(config.get(PROP_INDEX_PREFIX), PROP_INDEX_PREFIX_DEFAULT);
```
A general discussion - would it make sense to add a random String to this prefix ? Since we are providing this as a configurable property, technically there could be a case where 2 different tenants configure the same index prefix and we are not handling that here.
Appending a random String to the prefix would also help in the case if the tenant has not configured this property.

@fabriziofortino @thomasmueller

nit0906 commented on oak-benchmarks-elastic/src/main/java/org/apache/jackrabbit/oak/benchmark/ElasticFullTextWithGlobalIndexSearchTest.java:
```diff
@@ -54,8 +54,8 @@ public ElasticFullTextWithGlobalIndexSearchTest(File dump, boolean flat, boolean
-                            new ExtractedTextCache(10 * FileUtils.ONE_MB, 100));
-                    ElasticsearchIndexProvider indexProvider = new ElasticsearchIndexProvider(coordinate);
+                            new ExtractedTextCache(10 * FileUtils.ONE_MB, 100), null);
+                    ElasticsearchIndexProvider indexProvider = new ElasticsearchIndexProvider(coordinate, null);
```
It would be better if we actually use a prefix here instead of null, these benchmarks tests would be executed on a single ES server and would act as multiple tenants - so would make sense to pass a legit prefix here.

Also, adding some randomization to the prefix would be required (See comment in ElasticsearchIndexProviderService). 

Otherwise we would need some mechanism to delete the index from elastic server after each test run (I don't think currently we have a delete mechanism exposed in the client - if we do have, that could be another problem).

We can handle the cleanup externally too (shell script calling index delete curl req) - but I would like that to happen after the complete job is run (all the benchmark tests) - so I think it would be good to handle where multiple tests can run on the same ES server without having the need to delete the underlying index after every test run.

nit0906 commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:
```diff
@@ -31,6 +33,8 @@
 
 public class ElasticsearchIndexDefinition extends IndexDefinition {
 
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchIndexDefinition.class);
```
Is this needed ? I don't see this addition being used anywhere in the class.

nit0906 reviewed (commented):
Some suggestions and general discussion around the current strategy implemented.

fabriziofortino reviewed (commented):
`indexPrefix` is something we would need every time we connect to ES. Wouldn't it make sense (and probably simplify the code as well) to have this property as part of the `ElasticsearchConnection` object?
Moreover, it should be annotated with `@NotNull` to avoid conflicts. The benchmark tests should pass a different identifier (eg: current timestamp) every time they run instead of `null`.

Commit:
OAK-9004: Multitenant index name strategy with reindex support
* Passing indexPrefix through ElasticSearchConnection

Commit:
OAK-9004: Multitenant index name strategy with reindex support
* Removing unused field

fabriziofortino commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java:
```diff
@@ -85,7 +86,7 @@ public String getRemoteIndexName() {
     private String setupIndexName() {
         // TODO: implement advanced remote index name strategy that takes into account multiple tenants and re-index process
-        return getESSafeIndexName(getIndexPath() + "-" + getReindexCount());
+        return getESSafeIndexName(indexPrefix + getIndexPath() + "-" + getReindexCount());
```
should we add a `.` to separate `indexPrefix` from `indexPath`?

averma21 replied:
`getESSafeName` converts `/` to `_` and indexPath starts with `/` so these two are already separated. Eg if indexPrefix is `15Apr` and index name is `eIndexOne` then remote index name is `apr15_eindexone-1`

fabriziofortino replied:
it's OK then

fabriziofortino replied:
@averma21 I thought more about this. I think we should add a `.` anyway because `indexPath` starting with `/` is not always true.
Looking at the benchmarks, the final index name will be `benchmarkelasticglobal$ms`. We always need a fixed delimiter for the role permissions. Probably we will need additional validation on the prefix (eg: it cannot contains dots) but we can improve this in a separate PR.

fabriziofortino reviewed (approved):
We should probably need to revisit and add some validations to the prefix in the future to avoid problems whit the API key roles (https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html).
This can be done in a separate PR.

Commit:
OAK-9004: Multitenant index name strategy with reindex support
* Fixing exception in tests

nit0906 reviewed (approved):
LGTM.

thomasmueller commented on oak-benchmarks-elastic/src/main/java/org/apache/jackrabbit/oak/benchmark/ElasticFullTextWithGlobalIndexSearchTest.java:
```diff
@@ -40,7 +40,7 @@
 
     private ElasticsearchConnection coordinate;
-    private final String ELASTIC_GLOBAL_INDEX = "elasticGlobal";
+    private final String ELASTIC_GLOBAL_INDEX = "elasticGlobal" + System.nanoTime();
```
I suggest to use currentTimeMillis() instead of nanoTime(). nanoTime() starts at 0 when starting the process, so it might not be unique. currentTimeMillis() is the number of milliseconds since 1970, so (unless if the test is faster than 2 ms) is guaranteed to be unique.

thomasmueller commented on oak-benchmarks-elastic/src/main/java/org/apache/jackrabbit/oak/benchmark/ElasticPropertyFTIndexedContentAvailability.java:
```diff
@@ -66,7 +66,7 @@
     private String currentFixtureName;
     private ElasticsearchConnection coordinate;
-    private final String ELASTIC_GLOBAL_INDEX = "elasticGlobal";
+    private final String ELASTIC_GLOBAL_INDEX = "elasticGlobal" + System.nanoTime();
```
Same as above.

thomasmueller commented on oak-benchmarks-elastic/src/main/java/org/apache/jackrabbit/oak/benchmark/ElasticPropertyFTIndexedContentAvailability.java:
```diff
@@ -101,7 +101,7 @@ public Jcr customize(Oak oak) {
                             .with((new ElasticGlobalInitializer(ELASTIC_GLOBAL_INDEX, storageEnabled)).async())
                                     // the WikipediaImporter set a property `title`
-                            .with(new FullTextPropertyInitialiser("elasticTitle", of("title"),
+                            .with(new FullTextPropertyInitialiser("elasticTitle" + System.nanoTime(), of("title"),
```
Same as above.

thomasmueller commented on oak-benchmarks-elastic/src/main/java/org/apache/jackrabbit/oak/benchmark/ElasticPropertyFTSeparatedIndexedContentAvailability.java:
```diff
@@ -45,7 +45,7 @@
     private String currentFixtureName;
     private ElasticsearchConnection coordinate;
-    private final String ELASTIC_GLOBAL_INDEX = "elasticGlobal";
+    private final String ELASTIC_GLOBAL_INDEX = "elasticGlobal" + System.nanoTime();
```
OK it gets repetitive... Maybe use a utility method to construct unique names?

thomasmueller commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchConnection.java:
```diff
@@ -45,23 +45,25 @@
-    public ElasticsearchConnection(String scheme, String host, Integer port) {
-        if (scheme == null || host == null || port == null) {
+    public ElasticsearchConnection(String scheme, String host, Integer port, String indexPrefix) {
+        if (scheme == null || host == null || port == null || indexPrefix == null) {
```
The problem here is that if you get a NPE, you don't know why. Instead, could se use separate lines for each parameter, e.g. as in https://stackoverflow.com/questions/26393129/how-to-make-notnull-throw-runtime-exception

    Objects.requireNonNull(param);

thomasmueller commented on oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java:
```diff
@@ -77,7 +77,8 @@ protected ContentRepository createRepository() {
                 elastic.getContainerIpAddress(),
-                elastic.getMappedPort(ElasticsearchConnection.DEFAULT_PORT)
+                elastic.getMappedPort(ElasticsearchConnection.DEFAULT_PORT),
+                "" + System.nanoTime()
```
Same as above (currentTimeMillis)

Commit:
OAK-9004: Multitenant index name strategy with reindex support
* Adding . as indexPrefix delimeter
* using currentTimeMillis instead of nanoTime

Commit:
OAK-9004: Multitenant index name strategy with reindex support
* Removing unwanted outputs

averma21 commented:
I removed `System.nanoTime()` and added a `.` as separator for indexPrefix.
If it looks good, I can create and submit a patch for just the last two commits.
After moving the index name initialization to `createRepository` method of each of the tests I was able to use `System.currentTimeMillis()` . It was not possible use that at construction time because there is very little time difference between construction of these objects.

averma21 commented:
I checked with a local ES intance and benchmarks are passing. There are no index name related exceptions. 
[result4.txt](https://github.com/oak-indexing/jackrabbit-oak/files/4486816/result4.txt)

thomasmueller commented on oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchPropertyIndexTest.java:
```diff
@@ -77,7 +77,8 @@ protected ContentRepository createRepository() {
                 elastic.getContainerIpAddress(),
-                elastic.getMappedPort(ElasticsearchConnection.DEFAULT_PORT)
+                elastic.getMappedPort(ElasticsearchConnection.DEFAULT_PORT),
+                "" + System.currentTimeMillis()
```
getUniqueIndexName here as well?

averma21 replied:
This is a different bundle so we can't use that utility method from benchmark bundle here. If needed in future we can create one utility method here as well.

nit0906 replied:
I will do this as part of the overall refactoring of elastic tests under this issue https://issues.apache.org/jira/browse/OAK-9014

thomasmueller reviewed (approved):
Only one comment

Commit:
OAK-9004: Multitenant index name strategy with reindex support
Adding missing license to a file

nit0906 commented:
Changes merged in trunk for https://issues.apache.org/jira/browse/OAK-9004.
